File: core/energy.py
"""Energy and arousal system: finite energy budget, arousal modulation,
selective attention, flow states, stress responses, circadian rhythm."""

from __future__ import annotations
import numpy as np
from collections import deque

# Aspects prioritized during stress
STRESS_PRIORITY = {'self-regulation', 'situational_awareness', 'agency', 'body_awareness'}

# Clusters for flow state detection
FLOW_CLUSTERS = {
    'executive': ['metacognition', 'self-regulation', 'self-monitoring', 'goal-setting'],
    'reflective': ['introspection', 'reflection', 'temporal_awareness', 'moral_awareness'],
    'social': ['theory_of_mind', 'social_awareness', 'emotional_awareness', 'situational_awareness'],
    'identity': ['self-concept', 'self-esteem', 'self-recognition', 'self-efficacy'],
    'drive': ['motivation', 'agency', 'goal-setting', 'self-development'],
}


class EnergySystem:
    """Resource-constrained modulation layer.

    Returns modifiers for learning rate, noise amplitude, and per-aspect
    coupling each tick.
    """

    def __init__(self, aspects: list[str], resting_weights: dict[str, float], *,
                 max_energy: float = 100.0, attention_slots: int = 5,
                 circadian_period: int = 2000):
        self.aspects = list(aspects)
        self.n = len(aspects)
        self.resting = np.array([resting_weights.get(a, 0.0) for a in aspects])
        self._idx = {a: i for i, a in enumerate(aspects)}

        self.max_energy = max_energy
        self.energy = max_energy
        self.energy_regen_rate = 0.15
        self.energy_cost_scale = 0.02

        self.arousal = 0.7
        self.arousal_baseline = 0.7
        self.arousal_decay = 0.02
        self.arousal_min = 0.05
        self.arousal_max = 2.0

        self.attention_slots = attention_slots
        self.attention = np.full(self.n, 0.15)
        self._prev_weights = None

        self.flow_active: dict[str, bool] = {}
        self.flow_timer = {c: 0 for c in FLOW_CLUSTERS}
        self.flow_threshold = 15
        self.flow_range = (0.2, 0.85)
        self.flow_stability_band = 0.15

        self.stress_level = 0.0
        self.stress_decay = 0.03
        self.stress_threshold = 0.35

        self.circadian_period = circadian_period
        self.iteration = 0

    def step(self, weight_dict: dict[str, float]) -> dict:
        """Advance one tick. Returns {lr_modifier, noise_modifier, aspect_scales}."""
        weights = np.array([weight_dict.get(a, 0.0) for a in self.aspects])
        displacement = np.abs(weights - self.resting)

        roc = np.abs(weights - self._prev_weights) if self._prev_weights is not None \
            else np.zeros(self.n)
        self._prev_weights = weights.copy()

        self._update_energy(displacement)
        self._update_stress(displacement, roc)
        self._update_arousal()
        self._update_attention(displacement, roc)
        self._update_flow(weights)

        circadian = self._circadian_factor()
        energy_frac = self.energy / self.max_energy

        lr_mod = self.arousal * max(energy_frac, 0.1) * circadian

        noise_mod = self.arousal * circadian
        if self.stress_level > 0.3:
            noise_mod *= (1.0 + 0.5 * self.stress_level)
        if any(self.flow_active.get(c, False) for c in FLOW_CLUSTERS):
            noise_mod *= 0.7

        aspect_scales = {}
        for i, a in enumerate(self.aspects):
            scale = self.attention[i]
            if energy_frac < 0.3:
                scale *= 0.3 + 0.7 * (energy_frac / 0.3)
            if self.stress_level > 0.3 and a in STRESS_PRIORITY:
                scale *= (1.0 + self.stress_level)
            for cname, members in FLOW_CLUSTERS.items():
                if self.flow_active.get(cname, False) and a in members:
                    scale *= 1.3
                    break
            aspect_scales[a] = float(np.clip(scale, 0.05, 3.0))

        self.iteration += 1

        return {
            'lr_modifier': float(lr_mod),
            'noise_modifier': float(noise_mod),
            'aspect_scales': aspect_scales,
        }

    def get_state(self) -> dict:
        attended = [self.aspects[i] for i in np.argsort(self.attention)[-self.attention_slots:]]
        return {
            'energy': round(self.energy, 2),
            'energy_pct': round(100 * self.energy / self.max_energy, 1),
            'arousal': round(self.arousal, 4),
            'stress': round(self.stress_level, 4),
            'attended': attended,
            'flow_states': [c for c, v in self.flow_active.items() if v],
            'circadian': round(self._circadian_factor(), 3),
        }

    def inject_perturbation(self, magnitude=0.8):
        self.stress_level = min(1.0, self.stress_level + min(1.0, magnitude))
        self.arousal = min(self.arousal_max, self.arousal + min(1.0, magnitude) * 0.6)

    # ── Internal ──

    def _update_energy(self, displacement):
        cost = self.energy_cost_scale * np.sum(displacement ** 2) * (1.0 + self.stress_level)
        self.energy -= cost
        mean_disp = np.mean(displacement)
        if mean_disp < 0.25:
            self.energy += self.energy_regen_rate * (1.0 - mean_disp / 0.25)
        self.energy = float(np.clip(self.energy, 0.0, self.max_energy))

    def _update_arousal(self):
        target = self.arousal_baseline
        if self.stress_level > 0.2:
            target += self.stress_level * 0.8
        energy_frac = self.energy / self.max_energy
        if energy_frac < 0.3:
            target *= energy_frac / 0.3
        target *= self._circadian_factor()
        self.arousal += self.arousal_decay * (target - self.arousal)
        self.arousal = float(np.clip(self.arousal, self.arousal_min, self.arousal_max))

    def _update_attention(self, displacement, roc):
        salience = 0.6 * roc + 0.4 * displacement
        if self.stress_level > 0.3:
            for a in STRESS_PRIORITY:
                if a in self._idx:
                    salience[self._idx[a]] += self.stress_level * 0.5
        top = set(np.argsort(salience)[-self.attention_slots:])
        for i in range(self.n):
            if i in top:
                self.attention[i] += 0.15 * (1.0 - self.attention[i])
            else:
                self.attention[i] += 0.08 * (0.15 - self.attention[i])
        self.attention = np.clip(self.attention, 0.05, 1.0)

    def _update_stress(self, displacement, roc):
        perturbation = max(np.max(roc), np.max(displacement) * 0.3)
        if perturbation > self.stress_threshold:
            spike = min(1.0, (perturbation - self.stress_threshold) / self.stress_threshold)
            self.stress_level = min(1.0, self.stress_level + spike * 0.4)
        self.stress_level *= (1.0 - self.stress_decay)
        self.stress_level = float(np.clip(self.stress_level, 0.0, 1.0))

    def _update_flow(self, weights):
        for cname, members in FLOW_CLUSTERS.items():
            indices = [self._idx[a] for a in members if a in self._idx]
            if not indices:
                continue
            cluster_abs = np.abs(weights[indices] - self.resting[indices])
            lo, hi = self.flow_range
            in_range = np.all((cluster_abs >= lo) & (cluster_abs <= hi))
            stable = np.std(cluster_abs) < self.flow_stability_band
            if in_range and stable and self.stress_level < 0.3:
                self.flow_timer[cname] += 1
            else:
                self.flow_timer[cname] = max(0, self.flow_timer[cname] - 2)
            self.flow_active[cname] = self.flow_timer[cname] >= self.flow_threshold

    def _circadian_factor(self):
        if self.circadian_period <= 0:
            return 1.0
        phase = 2.0 * np.pi * self.iteration / self.circadian_period
        return 0.75 + 0.25 * np.cos(phase)

File: native/src/energy.rs
//! Energy and arousal system: finite energy budget, arousal modulation,
//! selective attention, flow states, stress responses, circadian rhythm.
//!
//! Mirrors core/energy.py exactly.

use std::f64::consts::TAU;

// Aspect names that receive priority salience during stress.
const STRESS_PRIORITY: &[&str] = &[
    "self-regulation",
    "situational_awareness",
    "agency",
    "body_awareness",
];

// Flow cluster definitions: (name, member aspect names).
const FLOW_CLUSTERS: &[(&str, &[&str])] = &[
    (
        "executive",
        &["metacognition", "self-regulation", "self-monitoring", "goal-setting"],
    ),
    (
        "reflective",
        &["introspection", "reflection", "temporal_awareness", "moral_awareness"],
    ),
    (
        "social",
        &["theory_of_mind", "social_awareness", "emotional_awareness", "situational_awareness"],
    ),
    (
        "identity",
        &["self-concept", "self-esteem", "self-recognition", "self-efficacy"],
    ),
    (
        "drive",
        &["motivation", "agency", "goal-setting", "self-development"],
    ),
];

const N_CLUSTERS: usize = 5;

pub struct EnergyState {
    pub energy: f64,
    pub energy_pct: f64,
    pub arousal: f64,
    pub stress: f64,
    pub attended: Vec<String>,    // aspect names currently attended
    pub flow_states: Vec<String>, // active flow cluster names
    pub circadian: f64,
    pub lr_modifier: f64,
    pub noise_modifier: f64,
    pub aspect_scales: Vec<f64>, // per-aspect coupling scale
}

pub struct EnergySystem {
    energy: f64,
    max_energy: f64,
    arousal: f64,
    arousal_baseline: f64,
    arousal_decay: f64,
    stress_level: f64,
    stress_decay: f64,
    stress_threshold: f64,
    attention: Vec<f64>,       // per-aspect attention weight
    attention_slots: usize,
    flow_timer: Vec<u32>,      // per-cluster timer
    flow_active: Vec<bool>,
    flow_threshold: u32,
    prev_weights: Option<Vec<f64>>,
    resting: Vec<f64>,
    iteration: u64,
    circadian_period: u32,
    n: usize,
    aspect_names: Vec<String>,

    // Resolved indices: per cluster, which aspect indices belong to it.
    cluster_indices: Vec<Vec<usize>>,
    // Per aspect: is it a stress-priority aspect?
    is_stress_priority: Vec<bool>,
    // Per aspect: which cluster indices include it (for flow boost).
    aspect_cluster_membership: Vec<Vec<usize>>,
}

impl EnergySystem {
    /// Construct from a list of aspect names, their resting weights, and tuning params.
    ///
    /// `resting_weights` is parallel to `aspects`.
    pub fn new(
        aspects: Vec<String>,
        resting_weights: &[f64],
        max_energy: f64,
        attention_slots: usize,
        circadian_period: u32,
    ) -> Self {
        let n = aspects.len();

        // Build a name→index map.
        let idx_of = |name: &str| aspects.iter().position(|a| a == name);

        // Resolve cluster membership indices.
        let cluster_indices: Vec<Vec<usize>> = FLOW_CLUSTERS
            .iter()
            .map(|(_cname, members)| {
                members.iter().filter_map(|m| idx_of(m)).collect()
            })
            .collect();

        // Per-aspect stress-priority flag.
        let is_stress_priority: Vec<bool> = aspects
            .iter()
            .map(|a| STRESS_PRIORITY.contains(&a.as_str()))
            .collect();

        // Per-aspect: list of cluster indices that include this aspect.
        let aspect_cluster_membership: Vec<Vec<usize>> = (0..n)
            .map(|ai| {
                cluster_indices
                    .iter()
                    .enumerate()
                    .filter(|(_ci, members)| members.contains(&ai))
                    .map(|(ci, _)| ci)
                    .collect()
            })
            .collect();

        let resting = resting_weights.to_vec();

        Self {
            energy: max_energy,
            max_energy,
            arousal: 0.7,
            arousal_baseline: 0.7,
            arousal_decay: 0.02,
            stress_level: 0.0,
            stress_decay: 0.03,
            stress_threshold: 0.35,
            attention: vec![0.15; n],
            attention_slots,
            flow_timer: vec![0u32; N_CLUSTERS],
            flow_active: vec![false; N_CLUSTERS],
            flow_threshold: 15,
            prev_weights: None,
            resting,
            iteration: 0,
            circadian_period,
            n,
            aspect_names: aspects,
            cluster_indices,
            is_stress_priority,
            aspect_cluster_membership,
        }
    }

    /// Advance one tick. Returns full EnergyState including modifiers.
    pub fn step(&mut self, weights: &[f64]) -> EnergyState {
        // Displacement from resting.
        let displacement: Vec<f64> = (0..self.n)
            .map(|i| (weights[i] - self.resting[i]).abs())
            .collect();

        // Rate of change since last tick.
        let roc: Vec<f64> = match &self.prev_weights {
            Some(prev) => (0..self.n).map(|i| (weights[i] - prev[i]).abs()).collect(),
            None => vec![0.0; self.n],
        };
        self.prev_weights = Some(weights.to_vec());

        self.update_energy(&displacement);
        self.update_stress(&displacement, &roc);
        self.update_arousal();
        self.update_attention(&displacement, &roc);
        self.update_flow(weights);

        let circadian = self.circadian_factor();
        let energy_frac = if self.max_energy > 0.0 {
            (self.energy / self.max_energy).clamp(0.0, 1.0)
        } else {
            0.0
        };

        let lr_mod = self.arousal * energy_frac.max(0.1) * circadian;

        let mut noise_mod = self.arousal * circadian;
        if self.stress_level > 0.3 {
            noise_mod *= 1.0 + 0.5 * self.stress_level;
        }
        if self.flow_active.iter().any(|&f| f) {
            noise_mod *= 0.7;
        }

        // Per-aspect scales.
        let mut aspect_scales = Vec::with_capacity(self.n);
        for i in 0..self.n {
            let mut scale = self.attention[i];
            if energy_frac < 0.3 {
                scale *= 0.3 + 0.7 * (energy_frac / 0.3);
            }
            if self.stress_level > 0.3 && self.is_stress_priority[i] {
                scale *= 1.0 + self.stress_level;
            }
            // Flow boost: first matching active cluster wins.
            for &ci in &self.aspect_cluster_membership[i] {
                if self.flow_active[ci] {
                    scale *= 1.3;
                    break;
                }
            }
            aspect_scales.push(scale.clamp(0.05, 3.0));
        }

        // Attended aspects: top-k by attention weight.
        let attended = self.top_attended();
        let flow_states: Vec<String> = FLOW_CLUSTERS
            .iter()
            .enumerate()
            .filter(|(ci, _)| self.flow_active[*ci])
            .map(|(_, (cname, _))| cname.to_string())
            .collect();

        self.iteration += 1;

        let energy_pct = if self.max_energy > 0.0 {
            (100.0 * self.energy / self.max_energy).clamp(0.0, 100.0)
        } else {
            0.0
        };

        EnergyState {
            energy: self.energy,
            energy_pct,
            arousal: self.arousal,
            stress: self.stress_level,
            attended,
            flow_states,
            circadian,
            lr_modifier: lr_mod,
            noise_modifier: noise_mod,
            aspect_scales,
        }
    }

    /// Spike stress and arousal by `magnitude`.
    pub fn inject_perturbation(&mut self, magnitude: f64) {
        self.stress_level = (self.stress_level + magnitude.min(1.0)).min(1.0);
        self.arousal = (self.arousal + magnitude.min(1.0) * 0.6).min(2.0);
    }

    /// `0.75 + 0.25 * cos(2π * iteration / period)`
    pub fn circadian_factor(&self) -> f64 {
        if self.circadian_period == 0 {
            return 1.0;
        }
        let phase = TAU * self.iteration as f64 / self.circadian_period as f64;
        0.75 + 0.25 * phase.cos()
    }

    // ── Internal helpers ──

    fn update_energy(&mut self, displacement: &[f64]) {
        let cost = 0.02 * displacement.iter().map(|d| d * d).sum::<f64>()
            * (1.0 + self.stress_level);
        self.energy -= cost;

        let mean_disp = displacement.iter().sum::<f64>() / self.n as f64;
        if mean_disp < 0.25 {
            self.energy += 0.15 * (1.0 - mean_disp / 0.25);
        }
        self.energy = self.energy.clamp(0.0, self.max_energy);
    }

    fn update_stress(&mut self, displacement: &[f64], roc: &[f64]) {
        let max_roc = roc.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        let max_disp = displacement.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        let perturbation = max_roc.max(max_disp * 0.3);
        if perturbation > self.stress_threshold {
            let spike =
                ((perturbation - self.stress_threshold) / self.stress_threshold).min(1.0);
            self.stress_level = (self.stress_level + spike * 0.4).min(1.0);
        }
        self.stress_level *= 1.0 - self.stress_decay;
        self.stress_level = self.stress_level.clamp(0.0, 1.0);
    }

    fn update_arousal(&mut self) {
        let mut target = self.arousal_baseline;
        if self.stress_level > 0.2 {
            target += self.stress_level * 0.8;
        }
        let energy_frac = if self.max_energy > 0.0 {
            (self.energy / self.max_energy).clamp(0.0, 1.0)
        } else {
            0.0
        };
        if energy_frac < 0.3 {
            target *= energy_frac / 0.3;
        }
        target *= self.circadian_factor();
        self.arousal += self.arousal_decay * (target - self.arousal);
        self.arousal = self.arousal.clamp(0.05, 2.0);
    }

    fn update_attention(&mut self, displacement: &[f64], roc: &[f64]) {
        let mut salience: Vec<f64> = (0..self.n)
            .map(|i| 0.6 * roc[i] + 0.4 * displacement[i])
            .collect();

        if self.stress_level > 0.3 {
            for (i, slot) in salience.iter_mut().enumerate().take(self.n) {
                if self.is_stress_priority[i] {
                    *slot += self.stress_level * 0.5;
                }
            }
        }

        // Indices of top-k by salience.
        let top = top_k_indices(&salience, self.attention_slots);

        for i in 0..self.n {
            if top.contains(&i) {
                self.attention[i] += 0.15 * (1.0 - self.attention[i]);
            } else {
                self.attention[i] += 0.08 * (0.15 - self.attention[i]);
            }
            self.attention[i] = self.attention[i].clamp(0.05, 1.0);
        }
    }

    fn update_flow(&mut self, weights: &[f64]) {
        for ci in 0..N_CLUSTERS {
            let indices = &self.cluster_indices[ci];
            if indices.is_empty() {
                continue;
            }
            // Displacement of cluster members from resting.
            let cluster_abs: Vec<f64> = indices
                .iter()
                .map(|&i| (weights[i] - self.resting[i]).abs())
                .collect();

            let in_range = cluster_abs.iter().all(|&v| (0.2..=0.85).contains(&v));
            let std_dev = std_dev(&cluster_abs);
            let stable = std_dev < 0.15;

            if in_range && stable && self.stress_level < 0.3 {
                self.flow_timer[ci] += 1;
            } else {
                self.flow_timer[ci] = self.flow_timer[ci].saturating_sub(2);
            }
            self.flow_active[ci] = self.flow_timer[ci] >= self.flow_threshold;
        }
    }

    /// Return the names of the top `attention_slots` attended aspects.
    fn top_attended(&self) -> Vec<String> {
        let indices = top_k_indices(&self.attention, self.attention_slots);
        indices.iter().map(|&i| self.aspect_names[i].clone()).collect()
    }
}

// ── Pure utilities ──

/// Population standard deviation of a slice.
fn std_dev(xs: &[f64]) -> f64 {
    if xs.len() <= 1 {
        return 0.0;
    }
    let mean = xs.iter().sum::<f64>() / xs.len() as f64;
    let var = xs.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / xs.len() as f64;
    var.max(0.0).sqrt()
}

/// Indices of the top-k largest values (unsorted; matches numpy argsort behaviour).
fn top_k_indices(vals: &[f64], k: usize) -> Vec<usize> {
    let k = k.min(vals.len());
    let mut indices: Vec<usize> = (0..vals.len()).collect();
    // Partial sort: bring the k largest to the end (mirrors numpy argsort[-k:]).
    indices.sort_unstable_by(|&a, &b| {
        vals[a].partial_cmp(&vals[b]).unwrap_or(std::cmp::Ordering::Equal)
    });
    indices[vals.len() - k..].to_vec()
}
